### PR TITLE
chore(STONEO11Y-102): Alerting rule for persistent volume issues

### PR DESCRIPTION
**Pull request by Kousalya1998**
- Created a new alerting rule for persistent volume issues.
- Added related unit test cases.

### CODE REVIEW

**Comment by Kousalya1998** [comment]
Hello, please ignore the PodsNotReady files. Since my previous PR hasn't been merged yet, I had to keep those files as well.

**Comment by yftacherzog** [comment]
Do you have a dependency here on the previous PR? I think we can include just the relevant commit here.

File: prometheus/base/alerting/prometheus.pod_alerts.yaml
Comment (yftacherzog) [inline prometheus/base/alerting/prometheus.pod_alerts.yaml:45]: 1. we need the pv name on the message as well.
2. I don't know if this statement is necessarily correct. it might be that the issue is different. I'd rephrase it saying that the pv is in the {{ $labels.phase }} for more than 5 minutes.

File: test/promql/tests/test_persistentvolumeissues_pods.yaml
Comment (yftacherzog) [inline test/promql/tests/test_persistentvolumeissues_pods.yaml:50]: make it 5 minutes

File: test/promql/tests/test_persistentvolumeissues_pods.yaml
Comment (yftacherzog) [inline test/promql/tests/test_persistentvolumeissues_pods.yaml:106]: those are 2 legitimate phases. It should not be alerting for neither.

**Review by yftacherzog — CHANGES_REQUESTED** [review]
Change the message the alert throw, fix the test cases and rebase the PR on top of the upstream main branch rather than on the previous PR.

**Comment by yftacherzog** [comment]
> Hello, please ignore the PodsNotReady files. Since my previous [PR (for PodsNotReady](https://github.com/redhat-appstudio/o11y/pull/32) has been approved) hasn't been merged yet, I had to keep those files as well.

I don't think those PRs are dependent on one another, apart from sharing the same file. Start the PR from the main branch and rebase once the other PR is merged.

File: prometheus/base/alerting/prometheus.pod_alerts.yaml
Comment (avi-biton) [reply prometheus/base/alerting/prometheus.pod_alerts.yaml:45]: I agree, when running the query `kube_persistentvolume_status_phase`, pod label is not a part of the labels that we get, so it is not usable here.
The query is about the PV, not the pod and it should be included in the message.
So maybe something like:  
"Persistent Volume {{ $labels.persistentvolume }} in namespace {{ $labels.namespace }} is in {{ $labels.phase }} phase for more than 5 minutes"

File: prometheus/base/alerting/prometheus.pod_alerts.yaml
Comment (yftacherzog) [inline prometheus/base/alerting/prometheus.pod_alerts.yaml:38]: I'm not sure whether pod alerts is the best location for this rule. I don't think this is a pod metric.

I think we should have it on another file dedicated for persistent volumes or something like that.

File: test/promql/tests/test_persistentvolumeissues_pods.yaml
Comment (avi-biton) [inline test/promql/tests/test_persistentvolumeissues_pods.yaml:38]: pod is irrelevant for this rule, it can be removed

File: prometheus/base/alerting/prometheus.pod_alerts.yaml
Comment (Kousalya1998) [reply prometheus/base/alerting/prometheus.pod_alerts.yaml:38]: Added and updated all the other comments you have made. Thanks @yftacherzog & @avi-biton

File: test/promql/tests/test_persistentvolumeissues_pods.yaml
Comment (Kousalya1998) [reply test/promql/tests/test_persistentvolumeissues_pods.yaml:38]: That makes sense. I have updated it to `persistentvolume` now.

**Review by hmariset — APPROVED** [review]
LGTM

**Review by yftacherzog — CHANGES_REQUESTED** [review]
You have 4 commits on the PR. You should have 1. Please rebase on top of upstream/main.

File: test/promql/tests/test_persistentvolumeissues.yaml
Comment (yftacherzog) [inline test/promql/tests/test_persistentvolumeissues.yaml:78]: this phase should not be alerting, even when it's more than 5 minutes.

File: test/promql/tests/test_persistentvolumeissues.yaml
Comment (yftacherzog) [inline test/promql/tests/test_persistentvolumeissues.yaml:37]: The comments throughout the PR are about pod, but they should be about pv

File: test/promql/tests/test_persistentvolumeissues.yaml
Comment (avi-biton) [inline test/promql/tests/test_persistentvolumeissues.yaml:66]:
```suggestion
      # Persistent Volume is in Pending state for less than 5 mins so it will not be alerted.
```

Comment (avi-biton) [reply test/promql/tests/test_persistentvolumeissues.yaml:66]: All comments should be update accordingly.

**Review by amisstea — CHANGES_REQUESTED** [review]
LGTM but please rebase so there's only one commit.

**Review by amisstea — APPROVED** [review]

**Review by hmariset — APPROVED** [review]
LGTM

File: test/promql/tests/test_persistentvolumeissues.yaml
Comment (yftacherzog) [inline test/promql/tests/test_persistentvolumeissues.yaml:10]:
```suggestion
  # Alerted cases:
  - interval: 1m
    input_series:
      # Persistent volume is in the Pending state so it will be alerted.
```

File: test/promql/tests/test_persistentvolumeissues.yaml
Comment (yftacherzog) [inline test/promql/tests/test_persistentvolumeissues.yaml:64]:
```suggestion
  # Not Alerted cases:
  - interval: 1m
    input_series:
```

File: test/promql/tests/test_persistentvolumeissues.yaml
Comment (yftacherzog) [inline test/promql/tests/test_persistentvolumeissues.yaml:82]: in minute 5, the pv is not on any state (everything is at zero). Was that the intention?

Comment (Kousalya1998) [reply test/promql/tests/test_persistentvolumeissues.yaml:82]: Yes, I initially tried it by giving `1x3 0` for `phase=Available` and `0x3 1` for `phase='Bound' which is not alerted.

Based on your previous comment, I thought that was how you wanted it to be in the 5th minute.

Comment (yftacherzog) [reply test/promql/tests/test_persistentvolumeissues.yaml:83]: not sure which comment is that, but we need to make sure that the combination of those series represents a valid state of the environment. What's the scenario here that we want to simulate?

File: test/promql/tests/test_persistentvolumeissues.yaml
Comment (yftacherzog) [inline test/promql/tests/test_persistentvolumeissues.yaml:80]: I now realize the pv and ns are the same with this scenario and the previous, so we're actually overriding the same series.

File: test/promql/tests/test_persistentvolumeissues.yaml
Comment (yftacherzog) [reply test/promql/tests/test_persistentvolumeissues.yaml:83]: ok. Just make sure the comment represents the collection of series

File: test/promql/tests/test_persistentvolumeissues.yaml
Comment (yftacherzog) [inline test/promql/tests/test_persistentvolumeissues.yaml:69]: this is similar to the other thread. We have minute 5 with all phases with zero, which I don't think is a valid combination.

File: test/promql/tests/test_persistentvolumeissues.yaml
Comment (yftacherzog) [reply test/promql/tests/test_persistentvolumeissues.yaml:80]: The combination needs to be unique - otherwise, it's the same time series.
Suppose that you have this time series:
```
'my_metric{l1="a", l2="x"}
```
To add another time series, we need either the metric or one of the labels key/value be different.

In this case, we need either different namespace or a differ pv (or both).

File: test/promql/tests/test_persistentvolumeissues.yaml
Comment (Kousalya1998) [reply test/promql/tests/test_persistentvolumeissues.yaml]: I removed this test case.

**Review by yftacherzog — COMMENTED** [review]
I think you're good. Please rebase and resolve the conflicts.

**Comment by Kousalya1998** [comment]
> I think you're good. Please rebase and resolve the conflicts.

I need your approval too and then only I can merge.

**Review by yftacherzog — APPROVED** [review]

**Pull request merged**
